Issue 4531: Use new BookKeeper API and Upgrade to BookKeeper 4.9.2

eolivelli opened the pull request:
**Change log description**  
BookKeeper 4.8 introduced a new API (org.apache.bookkeeper.client.api).
The old API, LedgerHandler, is going to be deprecated officially in next upcoming release (4.11).
The new API is based on CompletableFuture facility of the JDK.
The new API allows access to the internal refcounted Netty buffer used by BK client.

**Purpose of the change**  
Fixes #4531.

Switch from using LedgerHandle to WriteHandle and ReadHandle.
I am changing BK version to 4.9.2.

**What the code does**  
Upgrade BookKeeper to 4.9.2 in Java code (not docker images)
Replace  usages of LedgerHandle with WriteHandle and ReadHandle.
Use the new CompletableFuture based API.

**How to verify it**  
Tests should run without errors.

eolivelli commented:
@ravisharda this patch is the same as #4394 but with only one signed commit, this way DCO is happy.

RaulGracia commented:
Thanks for fixing DCO @eolivelli. Could you please have a look to the codecov check? The report shows some code lines that are not covered by any test; it would be great if you could add some tests cases to cover them.

eolivelli commented:
@RaulGracia 
from CodeCov report I see that the only relevant class changed by my patch is BookKeeperLog.java
https://codecov.io/gh/pravega/pravega/compare/560d085c0657c7e491286e77b3b5a447b9c1bd00...d045a0aebca1c25b7d1dcbeb360ba6349f597bd0/changes

but missed lines are mostly about "catch" blocks for special corner cases, in order to trigger them I should have a mock BookKeeper client that simulates errors during close operation of ledgers.
I don't this it is worth.

Do you see some important case to cover with a specific test case ?

RaulGracia commented:
@andreipaduroiu @fpj could you please do give one more look to this PR? At the moment:
- System tests are passing (3/3)
- Verifying that in terms of performance it does not introduce any issue (will let you know when the test completes).

andreipaduroiu commented on segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/LogReader.java:
```diff
@@ -75,12 +77,7 @@
-                    log.error("Unable to close LedgerHandle for Ledger {}.", this.currentLedger.handle.getId(), bkEx);
-                }
-
+                this.currentLedger.close();
```
Why did you remove the try-catch here but not below (at 237)?

eolivelli replied:
@andreipaduroiu 
because at line 237 we have
`Ledgers.close(xxx)`
that may throw an exception, and it is actually not possible that it happens

here we are using currentLedger.close(); (LogReader#close)) that calls exactly the code at line 237, and there is not need to add a catch clause.

andreipaduroiu commented:
This looks good, however codecov isn't very happy about the unit test coverage.

fpj commented on segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/Ledgers.java:
```diff
@@ -35,13 +41,27 @@
+        if (cause instanceof BKException) {
+            return (BKException) cause;
+        } else {
+            BKUnexpectedConditionException wrapper = new BKUnexpectedConditionException();
```
According to codecov, this path is not being exercised. Could we test this handler directly to exercise all code paths?

eolivelli replied:
sure, will do

eolivelli replied:
done

eolivelli commented on segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/DebugLogWrapper.java:
```diff
@@ -243,6 +230,18 @@ public boolean reconcileLedgers(List<LedgerHandle> candidateLedgers) throws Dura
         return changed;
     }
+    
+    private static long getHandleLength(Handle handle) {
```
it is because in theory this method can be called with a ReadHandle in a test, but actually it isn't
I feel it is safe to leave this form.
I will send a patch to BK in order to have getLength() available on Handle class

eolivelli replied:
I have double checked and WriteHandle is a subclass of ReadHandle. So code can be simpler

eolivelli replied:
Basically I have dropped this method

eolivelli commented:
@fpj the patch is ready for your final review.

RaulGracia commented:
@eolivelli looks like the build is failing now, which was not the case before the lasts commits, right?
```
io.pravega.segmentstore.storage.impl.bookkeeper.BookKeeperLogTests$RegularBookKeeperLogTests > testAppendTransientBookieFailure FAILED
    java.util.concurrent.CompletionException: io.pravega.common.util.RetriesExhaustedException: io.pravega.segmentstore.storage.WriteFailureException: Active Ledger is closed.
```

eolivelli commented:
I think that @andreipaduroiu already pointed out that it seems that with new BK client the client is observing a different series of state changes.
In 4.9 we refactored how the writer handles ledger metadata, and now the LedgerMetadata structure is immutable.

Maybe it can we a problem of the test.
Let me investigate this failure.

I did not check other commits that went into master branch since I have fixed the conflicts last time. This is also a good path to explore.

I will restart Travis build.

eolivelli commented:
@RaulGracia I don't have my laptop now. I will put the test in a loop as soon as I can.
Unfortunately on Travis it is not possible to see the logs

eolivelli commented:
This time I can't find any  coverage changes related to my patch.

eolivelli commented:
@RaulGracia 
running BookKeeperLog tests on my laptop I noticed this error:
```
io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 16777216 byte(s) of direct memory (used: 469762048, max: 477626368)
        at io.netty.util.internal.PlatformDependent.incrementMemoryCounter(PlatformDependent.java:652)
        at io.netty.util.internal.PlatformDependent.allocateDirectNoCleaner(PlatformDependent.java:606)
        at io.netty.buffer.PoolArena$DirectArena.allocateDirect(PoolArena.java:764)
        at io.netty.buffer.PoolArena$DirectArena.newChunk(PoolArena.java:740)
        at io.netty.buffer.PoolArena.allocateNormal(PoolArena.java:244)
        at io.netty.buffer.PoolArena.allocate(PoolArena.java:214)
        at io.netty.buffer.PoolArena.allocate(PoolArena.java:146)
        at io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:324)
        at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:185)
        at org.apache.bookkeeper.common.allocator.impl.ByteBufAllocatorImpl.newDirectBuffer(ByteBufAllocatorImpl.java:164)
        at org.apache.bookkeeper.common.allocator.impl.ByteBufAllocatorImpl.newDirectBuffer(ByteBufAllocatorImpl.java:158)
        at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:185)
        at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:176)
        at io.netty.channel.epoll.AbstractEpollChannel.newDirectBuffer0(AbstractEpollChannel.java:323)
        at io.netty.channel.epoll.AbstractEpollChannel.newDirectBuffer(AbstractEpollChannel.java:309)
        at io.netty.channel.epoll.AbstractEpollChannel.newDirectBuffer(AbstractEpollChannel.java:292)
        at io.netty.channel.epoll.AbstractEpollStreamChannel.filterOutboundMessage(AbstractEpollStreamChannel.java:533)
        at io.netty.channel.AbstractChannel$AbstractUnsafe.write(AbstractChannel.java:877)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.write(DefaultChannelPipeline.java:1391)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:738)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:730)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:816)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:723)
        at org.apache.bookkeeper.util.ByteBufList$Encoder.write(ByteBufList.java:328)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:738)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:730)
```
Relevant ticket on BK 4.9.x is https://github.com/apache/bookkeeper/pull/1754

An option to fix this is to increase memory amount given to Gradle in order to run tests

Do you have some kind of long running tests in order to see if we are hitting some kind of leak ? enabling Netty paranoic leak detection mode would help..

I will also double check our code.
Using the new BK API you have to explicitly release references to data that you read from BK, but this patch looks correct.

RaulGracia commented:
Thanks @eolivelli, yes it looks like you need to run the tests with more memory resources, but still 500MBs should be enough, right? Still, I don't know if this is related to the sporadic failures in test `testAppendTransientBookieFailure`.

By the way, @andreipaduroiu enabled some time back the Netty leak detector: https://github.com/pravega/pravega/pull/4750
I assume that this would tell us if the problem is related to memory leaks related to Netty buffers, right?

eolivelli commented:
@RaulGracia 
`Still, I don't know if this is related to the sporadic failures in test testAppendTransientBookieFailure.`
Yes I am investigating the cause of that failure. But it is not the only failing test

eolivelli commented:
Tests are failing randomly.
When I observe the error then the test (not  always the same one) fails

eolivelli commented:
@RaulGracia the problem is about DirectMemory and hot heap.

With this commit
https://github.com/pravega/pravega/pull/4798/commits/da1c5dc00542de383ece2519e06e2354e76c6858
I am setting 2g of DirectMemory and I cannot reproduce any more test failure on my laptop.
I could not find any other way to configure that value with Gradle (but I am still very new to Gradle), if you know a better way I am open to suggestions.

I had also checked BookKeeper client configuration that from 4.9 has a set of advanced ByteBuffer allocator configuration parameters, and if you tell to BK to prefer using the Heap tests are less flaky, but I think it is not a path worth to investigate more.

BK 4.9 and the new API in general prefer (Netty) direct memory and in my opinion this is the cause of flakyness.

In my opinion this patch is good to go

RaulGracia commented:
@eolivelli thanks for looking into the issue. Let's see for the final comments of @fpj on this PR.

RaulGracia commented:
@eolivelli I have also started a couple of complete builds with the last version of this PR. Will post the result once they complete.

eolivelli commented:
As said on slack I am not able to reproduce the failure locally.
I guess the problem is about a slow startup of the bookie.

The only kind of problems I have reproduced locally with this branch are about  ZK ConnectionLoss events/errors.

Off topic: We could make the tests more CI friendly by disabling fsyncs for the bookie and for ZK server. This won't help the restart of the bookie.

I don't have a local env to reproduce the issue (one very  slow machine), and I can't install VirtualBox

eolivelli commented:
@andreipaduroiu I have investigate the problems on Jenkins and I finally got to reproduce the error on an old laptop with 2 cores.

I added new log and finally I demonstrated that with a very slow machine the number of retries (3 with a timeout of 1000ms) ends **before starting the bookie again**. 

I have fixed the failure but increasing the BK timeout and number of max write attempts.

I have also disabled the journalSyncData flag in order to use less fdatasync on Bookies for tests, this should help Jenkins, but this is not strictly needed to help that test.

@RaulGracia  please confirm that Jenkins is not happy.
The patch should be good to go now

RaulGracia commented:
@eolivelli thanks, will try this one on Jenkins and see what happens. Also, we can execute several Travis builds as well to see if the issue is reproduced. Apart from that, if the `journalSyncData` flag change is not strictly needed, I would rather prefer to leave it as it was, as in Pravega we always have this set to true production.

eolivelli commented:
@RaulGracia reverted

eolivelli commented:
@fpj it looks like this patch is finally ready

andreipaduroiu commented:
Awesome.  Glad to see this coming through.
